File: web/src/types/workspace.ts
import type { BackendKind } from "./storage-backend";

export type WorkspaceMemberRole = "owner" | "writer" | "reader";

export interface Workspace {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  // Storage lives on catalogs now; these summarize the workspace's default
  // catalog and are null when the workspace has no catalog attached yet.
  default_catalog: string | null;
  storage_backend_id: string | null;
  storage_backend_kind: BackendKind | null;
  created_at: string;
}

export interface WorkspaceMember {
  workspace_id: string;
  user_id: string;
  role: WorkspaceMemberRole;
}

File: api/src/api/schemas/workspace.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class WorkspaceCreate(BaseModel):
    slug: str
    name: str


class WorkspaceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    name: str
    created_at: datetime
    # Storage is catalog-scoped now; these summarize the workspace's *default*
    # catalog so existing UI (the switcher's backend badge) keeps rendering.
    default_catalog: str | None = None
    storage_backend_id: uuid.UUID | None = None
    storage_backend_kind: str | None = None


class MemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    workspace_id: uuid.UUID
    user_id: uuid.UUID
    role: str


class AddMemberRequest(BaseModel):
    user_id: uuid.UUID
    role: str = "reader"


class AdminUserWorkspace(BaseModel):
    """A workspace and the target user's role in it (``None`` = not a member).

    Drives the admin "manage workspaces" view, which lists every workspace so an
    admin can grant or change a user's membership in one place.
    """

    workspace_id: uuid.UUID
    slug: str
    name: str
    role: str | None = None


class SetMembershipRequest(BaseModel):
    role: str
